common/tqid/task_queue_id.go

```diff
@@ -0,0 +1,339 @@
+		IsRoot() bool
+		IsSticky() bool
+		Kind() enumspb.TaskQueueKind
+
+		// RpcName returns the mangled name of the task queue partition, to be used in RPCs.
+		//
+		// RPC names look like this:
+		//
+		//  sticky partition:			<sticky name>
+		//	root normal partition: 		<task queue name>
+		//	non-root normal partition: 	/_sys/<task queue name>/<partition id>
+		//
+		// This scheme lets users use anything they like for a base name, except for strings
+		// starting with "/_sys/", without ambiguity.
+		//
+		// For backward compatibility, unversioned low-level task queues with partition 0 do not
+		// use mangled names, they use the bare base name.
+		RpcName() string
+		Key() PartitionKey
+	}
+
+	// NormalPartition is used to distribute load of a TaskQueue in multiple Matching instances. A normal partition is
+	// identified by `partitionId`. The partition with ID 0 is called a root partition.
+	NormalPartition struct {
+		taskQueue   *TaskQueue
+		partitionId int
+	}
+
+	// StickyPartition is made by SDK for a single workflow worker to keep workflow tasks of the same execution
+	// in the same worker for caching benefits. Each sticky partition is identified by a unique `stickyName`
+	// generated by SDK. A StickyPartition can only have workflow task type.
+	StickyPartition struct {
+		stickyName string
+		taskQueue  *TaskQueue
+	}
+
+	// PartitionKey uniquely identifies a task queue partition, to be used in maps.
+	// Note that task queue kind (sticky vs normal) and normal name for sticky task queues are not
+	// part of the task queue partition identity.
+	PartitionKey struct {
+		namespaceID string
+		name        string
+		partitionId int
+		taskType    enumspb.TaskQueueType
+	}
+)
+
+var _ Partition = (*NormalPartition)(nil)
+var _ Partition = (*StickyPartition)(nil)
+
+var (
+	ErrNoParent      = errors.New("root task queue partition has no parent")
+	ErrInvalidDegree = errors.New("invalid task queue partition branching degree")
+	ErrNonZeroSticky = errors.New("only sticky partitions can not have non-zero partition ID")
+)
+
+// FromFamilyName takes a user-provided task queue name (aka family name) and returns a TaskQueueFamily. Returns an
+// error if name looks like a mangled name.
+func FromFamilyName(namespaceId string, name string) (*TaskQueueFamily, error) {
+	if strings.HasPrefix(name, nonRootPartitionPrefix) {
+		return nil, serviceerror.NewInvalidArgument("task queue family name cannot have prefix /_sys/ " + name)
+	}
+	return &TaskQueueFamily{
+		namespaceId: namespace.ID(namespaceId),
+		name:        name,
+	}, nil
+}
+
+// UnsafeTaskQueueFamily should be avoided as much as possible. Use FromFamilyName instead as it validates the tq name.
+func UnsafeTaskQueueFamily(namespaceId string, name string) *TaskQueueFamily {
+	return &TaskQueueFamily{namespace.ID(namespaceId), name}
+}
+
+func FromProto(proto *taskqueuepb.TaskQueue, namespaceId string, taskType enumspb.TaskQueueType) (Partition, error) {
```

Comment:
Same as with TaskQueueFamily. We should incorporate `Partition` into function name. It sould be `PartitionFromProto` or something like this.